New Resources: aws_acm_certificate and aws_acm_certificate_validation

flosell opened the pull request:
This is a second draft trying to implement automated ACM certificate issuing based on my suggestions in #2418. It's similar to PR #2801 but uses a resource instead of a data source to implement polling for certificate issuing as a data source implementation seems to have a few pitfalls for the users. 

Here is what it does: 

* Add a `aws_acm_certificate` resource that requests a new certificate and returns once `validationOptions` (the information what needs to be done to validate ownership of the domain) are available (for some reason, those aren't immediately available). Only supports DNS validation at this point
* Add a `aws_acm_certificate_validation` resource that does some basic sanity checks and then waits for the given certificate to be issued
* Add an acceptance test that runs the whole certificate issuing flow (using route53 for DNS validation)

Current Limitations: 

* Only supports DNS validation: E-Mail validation is hard to completely automate and test (and already discussed at length in hashicorp/terraform#4782) so I did not invest time here. It shouldn't be a problem to add this later if there is demand for it
* Only supports certificates that are` AMAZON_ISSUED`: Importing existing certificate data could be added later if there is demand for it

Test considerations: 

Acceptance Tests need a publicly resolvable Route53 zone to make certificate validation work. To run tests, set the `ACM_CERTIFICATE_ROOT_DOMAIN` environment variable to the domain name of the Route53 zone. The test will use data-providers to look up the zone-id and add CNAMES to allow ACM to validate ownership of the domain: 

```
ACM_CERTIFICATE_ROOT_DOMAIN="sandbox.sellmayr.net" make testacc TEST=./aws TESTARGS='-run=TestAccAwsAcm'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAwsAcm -timeout 120m
=== RUN   TestAccAwsAcmCertificateDataSource_noMatchReturnsError
--- PASS: TestAccAwsAcmCertificateDataSource_noMatchReturnsError (20.45s)
=== RUN   TestAccAwsAcmResource_certificateIssuingFlow
--- PASS: TestAccAwsAcmResource_certificateIssuingFlow (324.23s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	344.721s
```

Here are a few things I'm unsure about: 

* Is it a good idea to model waiting for certificate issuing in a resource? I feel it's harder for the user to shoot himself in the foot this way but it's also a bit unintuitive to have a resource that doesn't correspond to any resource in AWS, especially when destroying it: Does destroying a validation destroy the certificate? It doesn't but is that obvious to the user? 
* AWS does not expect `subject_alternative_names` to contain values when requesting a certificate but `DescribeCertificate` always includes the `domain_name` in `subject_alternative_names`. I'm currently filtering this so terraform doesn't detect this as a change. Is there a better way to do this?  
* Should we have the `validation_method` property? It currently only allows `DNS` as a value so we could do without. However, it probably makes adding new validation methods easier

flosell commented:
Removed the `[WIP]` flag, from my point of view this PR is ready for review.

pawelsocha commented:
@bflad what you think? it's possible to merge this PR?

xyloman commented:
Agreed we have a new terraform that would benefit from this.

bflad commented:
You caught me. 😄  I'm in the ACM neighborhood right now with #1837 and was actually planning on looking at this right after so I didn't have too much context switching. While that one is waiting for some feedback let me at least provide an initial review for this one. More shortly.

bflad commented on website/docs/r/acm_certificate_validation.html.markdown:
````diff
@@ -0,0 +1,65 @@
+```hcl
+resource "aws_acm_certificate" "cert" {
+    domain_name = "example.com"
+	validation_method = "DNS"
````
Nitpick: formatting

flosell replied:
tabs and spaces... 😄 Fixed.

flosell commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -0,0 +1,237 @@
+package aws
+
+import (
```
Done, thanks for the pointers to `goimports`, that helps a lot!

flosell commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -0,0 +1,237 @@
+		},
+
+		Schema: map[string]*schema.Schema{
+			"domain_name": &schema.Schema{
```
Fixed.

flosell commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -0,0 +1,237 @@
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"domain_name": &schema.Schema{
```
Fixed.

flosell commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -0,0 +1,237 @@
+	acmconn := meta.(*AWSClient).acmconn
+	params := &acm.RequestCertificateInput{
+		DomainName:       aws.String(d.Get("domain_name").(string)),
+		ValidationMethod: aws.String("DNS"),
```
Fixed.

flosell commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -0,0 +1,237 @@
+func resourceAwsAcmCertificateDelete(d *schema.ResourceData, meta interface{}) error {
+	acmconn := meta.(*AWSClient).acmconn
+
+	if err := resourceAwsAcmCertificateRead(d, meta); err != nil {
```
fixed.

flosell replied:
Hmm, I probably stole that from the `eip` resource... I guess it's not really necessary and I fixed it in the way you outlined, which gave me another thought:
The point of this is to tolerate cases where certificates were already deleted outside of terraform, correct? Wouldn't it make more sense to catch `ErrCodeResourceNotFoundException` in the read-operation instead?

bflad replied:
We actually do that in a bunch of resources as well. Probably good to add into this one for good measure too 👍

bflad replied:
Oh and by that I mean in both read and delete. Some folks when they get stuck in a really bad configuration for `terraform destroy` will run the (not advised) `-refresh=false`, that's why we try to add it to both read and delete.

flosell replied:
Interesting, makes sense. Added the check to read in both resources as well.

flosell commented on aws/resource_aws_acm_certificate_test.go:
```diff
@@ -0,0 +1,311 @@
+		}
+
+		// Verify the error is what we want
+		acmerr, ok := err.(awserr.Error)
```
Fixed, thanks for the pointer to `isAWSErr`!

flosell commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -0,0 +1,237 @@
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
```
Using the constant now. 

Supporting E-Mail validation will probably add a bit more work and complexity. It's also hard to test. 
At a minimum, we'd need to consider cases where domain validation information isn't present to make sure the provider doesn't break if someone tries to use it with `EMAIL`. 

Since this PR is already pretty big, my initial idea was to add support for E-Mail validation in a separate PR later if there is demand for it. If DNS validation works smoothly, I'd guess most people (unless they can't for some reason) will use it and not bother with E-Mail.

Does that make sense? If you prefer adding E-Mail support right now, I'd also be willing to invest a bit of time to get it in.

flosell commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -0,0 +1,237 @@
+					return
+				},
+			},
+			"certificate_arn": &schema.Schema{
```
* Renamed to `arn` in `acm_certificate`. I left `certificate_arn` in `acm_certificate_validation` for now as I feel it's not quite as obvious which `arn` is meant there. No strong opinion though, happy to change it if you feel otherwise
* Removed `ForceNew`

flosell commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -0,0 +1,237 @@
+			return resource.NonRetryableError(fmt.Errorf("Error describing certificate: %s", err))
+		}
+
+		if *resp.Certificate.Type != "AMAZON_ISSUED" {
```
Mainly because I haven't tested it yet. I'm guessing `DomainValidationOptions` would be missing so similar to support for E-Mail validation we'd need to be more flexible with what data we expect from the API. 

For now, I fixed the the nitpick. For the rest, just like E-Mail validation, we have to make the decision to split work on this into a separate PR or work on it in this one.

flosell replied:
Done, removed the limitation on `AMAZON_ISSUED` certificates, `IMPORTED` certificates. 

Also added a warning to the docs that this should be used carefully as it might lead to fragile terraform projects: `IMPORTED` certificates can be imported and managed by terraform but not re-created after being destroyed.

flosell commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -0,0 +1,237 @@
+			return resource.NonRetryableError(fmt.Errorf("Certificate has type %s, only AMAZON_ISSUED is supported at the moment", *resp.Certificate.Type))
+		}
+
+		if err := d.Set("domain_name", resp.Certificate.DomainName); err != nil {
```
Fixed, I guess that was me getting paranoid after I missed some errors on other `d.Set()` calls 😄

flosell commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -0,0 +1,237 @@
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
```
I tried it out, looks like this leads to some interesting interactions: When using the fully automatic validation flow (e.g. with route53 and the `acm_certificate_validation` resource), the `acm_certificate` resource completes first (with a `PENDING_VALIDATION` status), then the route53 record gets added and `acm_certificate_validation` waits until ACM issues the certificate. Since the `acm_certificate` resource isn't refreshed at that point, it stays in status `PENDING_VALIDATION`. After the next terraform run, this state switches to `ISSUED`. 

I haven't seen an immediate impact but I'd guess this can lead to some interesting surprises down the road :)

flosell commented on aws/resource_aws_acm_certificate_test.go:
```diff
@@ -0,0 +1,311 @@
+					testAccCheckTagsACM(&tags.Tags, "Foo", "Baz"),
+				),
+			},
+			// Test that validation times out if certificate can't be validated
```
I agree in principle, there's a catch though: Certificate requests have a [limit](https://docs.aws.amazon.com/acm/latest/userguide/acm-limits.html) (twice the number of ACM certificates per region per year so 200 by default but maybe a lot lower) that you hit quickly if you run a lot of tests. It can be increased with approval from the service team but I'm not sure by how much (I got 400 requests per region now).

If we want to run these tests in CI, the HashiCorp account possibly needs a pretty high limit and every test that requests a certificate increases it.

flosell commented on aws/resource_aws_acm_certificate_test.go:
```diff
@@ -0,0 +1,311 @@
+				Config: testAccAcmCertificateConfig(domain, sanDomain),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAcmCertificateExists("aws_acm_certificate.cert", &conf, &tags),
+					testAccCheckAcmCertificateAttributes("aws_acm_certificate.cert", &conf, domain, sanDomain, "PENDING_VALIDATION"),
```
I found this pattern in some other tests, e.g. `testAccCheckAWSENIAttributes`. 
I guess it's a bit more paranoid in the sense that it checks the actual state in AWS without relying on the read operation. In theory, a no-op create and read would also result in the correct resource attributes in the state since they are initially populated from the resource declaration in the TF config, right? 

Is that worth the extra complexity? I'm not sure...
For now I added the checks you suggested, they definitely make the the expectations more explicit.

bflad replied:
Generally speaking we rely on the resource state checks (such as `resource.TestCheckResourceAttr`) as the acceptance testing is already paranoid about refreshing the resources from AWS. For many AWS services, having the _additional_ checks makes the acceptance testing more susceptible to eventual consistency issues. Some of the older provider testing code may have predated the state helpers or just was never converted over.

I would really consider getting rid of these unless there's some reason to keep them, which generally means an attribute should be added.

flosell replied:
Done, removed the custom checks. I don't think there's anything missing except the check that the certificate is in state `ISSUED` but the validation resource checks that.

flosell commented on aws/resource_aws_acm_certificate_test.go:
```diff
@@ -0,0 +1,311 @@
+	}
+
+	root_zone_domain := os.Getenv("ACM_CERTIFICATE_ROOT_DOMAIN")
+	domain := fmt.Sprintf("certtest.%s", root_zone_domain)
```
Done.

flosell commented on website/docs/r/acm_certificate.html.markdown:
```diff
@@ -0,0 +1,87 @@
+
+## Attributes Reference
+
+The following attributes are exported:
```
Fixed. I also removed the attributes reference section from the `acm_certificate_validation` docs since that resource doesn't export any additional attributes.

flosell commented on website/docs/r/acm_certificate.html.markdown:
```diff
@@ -0,0 +1,87 @@
+  }
+}
+
+data "aws_route53_zone" "zone" {
```
Removed. If I understand you correctly we leave the more complete example in the docs for validation? I think otherwise it will be hard to follow how these resources work together (e.g. that you should use the `certificate_arn` from the `aws_acm_certificate_validation` resource and not `aws_acm_certificate` if you want to make sure you depend on an issued certificate).

flosell commented:
@bflad Thanks for reviewing! I addressed most of your feedback and left comments with my thinking on the others. I'd be happy for a second opinion on those but I might play around with them a bit if I find time on the weekend.

joelhandwell commented:
Got this branch built and ran ```terraform plan``` with following code:

```hcl
locals {
  domains = [
    {domain = "domain1.com" zone_id = "Z10NYRJM6KIABC" instance_id = "i-0bb0bccfe3c31cabc"},
    {domain = "domain2.com" zone_id = "Z1ANQG9R5VAABC" instance_id = "i-0bb0bccfe3c31cabc" }
  ]
}

resource "aws_acm_certificate" "cert" {
  count = "${length(local.domains)}"
  domain_name = "${lookup(local.domains[count.index], "domain")}"
  subject_alternative_names = ["www.${lookup(local.domains[count.index], "domain")}"]
  validation_method = "DNS"
}

resource "aws_route53_record" "cert_validation" {
  count =  "${length(local.domains)}"

  name = "${element(aws_acm_certificate.cert.*.domain_validation_options.0.resource_record_name, count.index)}"
  type = "CNAME" 
  zone_id = "${lookup(local.domains[count.index], "zone_id")}"
  records = ["a"]  
  ttl = 60
}
```

And getting following error:
```
> terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.aws_route53_zone.zone[0]: Refreshing state...
data.aws_route53_zone.zone[1]: Refreshing state...

------------------------------------------------------------------------

Error: Error running plan: 1 error(s) occurred:

* aws_route53_record.cert_validation: 2 error(s) occurred:

* aws_route53_record.cert_validation[0]: Resource 'aws_acm_certificate.cert' does not have attribute 'domain_validation_options.0.resource_record_name' for variable 'aws_acm_certificate.cert.*.domain_validation_options.0.resource_record_name'
* aws_route53_record.cert_validation[1]: Resource 'aws_acm_certificate.cert' does not have attribute 'domain_validation_options.0.resource_record_name' for variable 'aws_acm_certificate.cert.*.domain_validation_options.0.resource_record_name'

```
Any idea if I'm misusing? 

And I confirmed that this works:
```
name = "${aws_acm_certificate.cert.0.domain_validation_options.0.resource_record_name}"
```

So I suspected if it's related with something like https://github.com/hashicorp/terraform/pull/10657

bflad commented on aws/resource_aws_acm_certificate_test.go:
```diff
@@ -0,0 +1,324 @@
+					resource.TestCheckResourceAttr("aws_acm_certificate.cert", "tags.Foo", "Baz"),
+				),
+			},
+			// Test that validation times out if certificate can't be validated
```
Please separate any of the validation testing into `resource_aws_acm_certificate_validation_test.go`

flosell replied:
Done.

bflad commented on aws/resource_aws_acm_certificate_test.go:
```diff
@@ -0,0 +1,324 @@
+  subject_alternative_names = ["%s"]
+
+  tags {
+    "Hello" = "World"
```
Nitpick: we can save the extra `testAccAcmCertificateConfigWithChangedTags` by parameterizing the tag names/values.

flosell replied:
Done.

bflad commented on aws/resource_aws_acm_certificate_test.go:
```diff
@@ -0,0 +1,324 @@
+
+func testAccAcmCertificateWithValidationConfig(domain string, sanDomain string) string {
+	return fmt.Sprintf(`
+resource "aws_acm_certificate" "cert" {
```
Nitpick: this can be `%s` from `testAccAcmCertificateConfig()`

flosell replied:
Done.

bflad commented on aws/resource_aws_acm_certificate_validation.go:
```diff
@@ -0,0 +1,150 @@
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+			"timeout": {
```
Generally speaking, we should probably use the built-in schema timeout functionality instead of our own attribute for consistency with other resources and handling validation.

```
&schema.Resource{
  # ... other config ...
  Timeouts: &schema.ResourceTimeout{
    Create: schema.DefaultTimeout(45 * time.Minute),
  }
}
# usage
resource.Retry(d.Timeout(schema.TimeoutCreate), func() ...)
```

flosell replied:
Done.

flosell commented:
@joelhandwell What you are trying to do looks reasonable to me. Unfortunately, it also reminds me of other issues I've had where my research always ended with "HCL isn't expressive enough to do this"... To me it doesn't look like something that could be addressed at the provider level (happy to be corrected on both statements). 

@bflad just skimmed through the rest of your feedback, makes sense to me. I hope I'll find some time over the next few days to work on this. I'll keep you posted.

flosell commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -0,0 +1,227 @@
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					if v.(string) != acm.ValidationMethodDns {
```
Done.

flosell commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -0,0 +1,227 @@
+	acmconn := meta.(*AWSClient).acmconn
+	params := &acm.RequestCertificateInput{
+		DomainName:       aws.String(d.Get("domain_name").(string)),
+		ValidationMethod: aws.String(acm.ValidationMethodDns),
```
Done.

flosell commented on aws/resource_aws_acm_certificate.go:
```diff
@@ -0,0 +1,227 @@
+	return resource.Retry(time.Duration(1)*time.Minute, func() *resource.RetryError {
+		resp, err := acmconn.DescribeCertificate(params)
+
+		if err != nil && isAWSErr(err, acm.ErrCodeResourceNotFoundException, "") {
```
Done.

flosell commented on aws/resource_aws_acm_certificate_test.go:
```diff
@@ -0,0 +1,311 @@
+	}
+}
+
+func testAccCheckAcmCertificateAttributes(n string, cert *acm.DescribeCertificateOutput, domain string, sanDomain string, status string) resource.TestCheckFunc {
```
Done.

flosell commented on website/docs/r/acm_certificate.html.markdown:
```diff
@@ -0,0 +1,87 @@
+
+* `domain_name` - (Required) A domain name for which the certificate should be issued
+* `subject_alternative_names` - (Optional) A list of domains that should be SANs in the issued certificate
+* `validation_method` - (Required) Which method to use for validation (only `DNS` is supported at the moment)
```
Done.

flosell commented:
@bflad unless I missed something this should be ready for round two: 

* Now Supporting E-Mail Validation and and importing certificates that aren't `AMAZON_ISSUED`
* Cleaned up code and tests

malaikah commented:
This will be a massive help to us.

bflad commented:
This has been released in [terraform-provider-aws version 1.9.0](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#190-february-09-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

daveadams commented:
Hello, this is great! The example code works fine. But I have one problem. How do I handle creating multiple DNS records for an ACM cert with multiple subject alternative names? All of the facts I need are embedded in the resource's `domain_validation_options` property, but the values in the maps inside that list are not directly accessible as far as I can tell. Example code:

    resource "aws_acm_certificate" "cert" {
      domain_name = "myname.example.com"
      subject_alternative_names = [
        "my-alt-name.example.com",
        "my-other-name.example.com",
      ]
      validation_method = "DNS"
    }

    output "validation-options" {
      value = "${aws_acm_certificate.cert.domain_validation_options}"
    }

This results in this output:

    validation-options = [
        {
            domain_name = myname.example.com,
            resource_record_name = _8b1eece1bd9eb24c465d20ee5e902e34.myname.example.com.,
            resource_record_type = CNAME,
            resource_record_value = _7fcc0e9fce650cfc93cf84e9b3afa0a8.acm-validations.aws.
        },
        {
            domain_name = my-alt-name.example.com,
            resource_record_name = _117f0bed1655685f817791a0d446fbaf.my-alt-name.example.com.,
            resource_record_type = CNAME,
            resource_record_value = _eae8c229110499e4c535fde67c6cd8c0.acm-validations.aws.
        },
        {
            domain_name = my-other-name.example.com,
            resource_record_name = _456f43b3eacb2496165b136bda0f47db.my-other-name.example.com.,
            resource_record_type = CNAME,
            resource_record_value = _61efde25808a9a12bbd24ac29ca0308c.acm-validations.aws.
        }
    ]

Fair enough. But then I am stuck at how to create those three records. My first instinct is to treat this property like a multi-instance resource, and do something like:

    resource "aws_route53_record" "cert_validation" {
      count = "${length(aws_acm_certificate.cert.domain_validation_options)}"
      name    = "${element(aws_acm_certificate.cert.domain_validation_options.*.resource_record_name, count.index)}"
      type    = "${element(aws_acm_certificate.cert.domain_validation_options.*.resource_record_type, count.index)}"
      records = ["${element(aws_acm_certificate.cert.domain_validation_options.*.resource_record_value, count.index)}"]
      zone_id = "${data.aws_route53_zone.zone.id}"
      ttl     = 300
    }

But that gives me an error: "value of 'count' cannot be computed"

If I tweak things around or even hardcode "count" to the correct value, then I get errors like: "Resource 'aws_acm_certificate.cert' does not have attribute 'domain_validation_options.*.resource_record_type'".

I've tried using locals as interstitial calculators, but I can't figure out how to get, say, an array of just the resource_record_name values, given the cert_validation structure.

I'm also not able to get this to work using just straight local variables instead of the resource attribute. There just don't seem to be sufficient tools in the interpolation language to manipulate this particular data structure. Not sure if it's easier to fix the resource or the interpolation language.

laurentr commented:
This exactly describes the challenge I was experiencing yesterday. I know you can attach multiple ACM certs to an ALB, but it's nice if I can minimize the number of certificate lifecycles to manage.

jnatten commented:
Has anyone been able to create multiple `aws_route53_record`'s for certificates that has multiple domains?

I am unable to dynamically lookup the `domain_validation_options`
If i try to access `aws_acm_certificate_cert.domain_validation_options.0.resource_record_name` i get the expected value. 

But if i try to access it dynamically in a resource block with count like this:
```
resource "aws_route53_record" "cert_validation" {
  count = "${length(local.certificate_additional_domains) + 1}" // One record for each additional domain + one for the main domain
  name = "${element(aws_acm_certificate.cert.domain_validation_options.*.resource_record_name, count.index)}"
  [ ... ]
}
```

`terraform plan` tells me that: 
```
* module.mymodule.aws_route53_record.cert_validation: At column 3, line 1: element: argument 1 should be type list, got type string in:

${element(aws_acm_certificate.cert.domain_validation_options.*.resource_record_type, count.index)})
```

If i try to output just `aws_acm_certificate.cert.domain_validation_options.*.resource_record_name` somewhere the result is `1`

daveadams commented:
I've found that it works if you use:

    name = "${lookup(aws_acm_certificate.cert.domain_validation_options[count.index], "resource_record_name")}"

And similar for the other fields.

jnatten commented:
@daveadams That works perfectly, thanks!

digitalkaoz commented:
it doesnt work i you use `count`s:

```hcl
variable "apps" {
  type    = "list"
  default = ["foo", "bar"]
}

# dns validation record
resource "aws_route53_record" "live_cert_validation" {
  count   = "${length(var.apps)}"
  name    = "${lookup(element(aws_acm_certificate.live.*.domain_validation_options[count.index], 0), "resource_record_name")}"
  type    = "${lookup(element(aws_acm_certificate.live.*.domain_validation_options[count.index], 0), "resource_record_type")}"
  zone_id = "${data.aws_route53_zone.live.*.id[count.index]}"
  records = ["${lookup(element(aws_acm_certificate.live.*.domain_validation_options[count.index], 0), "resource_record_value")}"]
  ttl     = 60
}
```

more weird: it works if i create the certificate in one stack, and the certificate dns records and the validation in the next stack. but all in one stack gives me the following error:

```
* aws_route53_record.live_cert_validation: At column 3, line 1: lookup: argument 1 should be type map, got type string in:

${lookup(element(aws_acm_certificate.live.*.domain_validation_options[count.index], 0), "resource_record_name")}
```

it tried it in a lot of obscure ways, but nothing seems to work, i guess the interpolation comes in play to early here...

@flosell @bflad any ideas?

daveadams commented:
Looks like you have a second "type =" in your route53 definition. You
should just have the one pointing to the domain_validation_options record
type attribute. I'm guessing it's getting created as the wrong type.

On Wed, Feb 28, 2018 at 11:29 AM, Brian Auron <notifications@github.com>
wrote:

> Hello, I'm trying to get this to work for me in (what I think is) a pretty
> simple manner:
>
> I've created a domain, a certificate request, and a validation request. I
> can see that the CNAME exists as expected, but the certificate request just
> sits in Pending Validation status for 72 hours and times out:
>
> resource "aws_acm_certificate" "nexus_certificate" {
>   domain_name = "${var.module_name}.${var.route53_zone_name}"
>   validation_method = "DNS"
>   tags {
>     Environment = "${var.module_name}-${var.route53_zone_name}"
>   }
> }
>
> resource "aws_acm_certificate_validation" "nexus_certificate_validation"{
>   certificate_arn = "${aws_acm_certificate.nexus_certificate.arn}"
>   validation_record_fqdns = ["${aws_route53_record.nexus_cert_validation_record.fqdn}"]
> }
>
> resource "aws_route53_record" "nexus_cert_validation_record" {
>   name    = "${aws_acm_certificate.nexus_certificate.domain_validation_options.0.resource_record_name}"
>   type    = "${aws_acm_certificate.nexus_certificate.domain_validation_options.0.resource_record_type}"
>   zone_id = "${var.route53_zone_id}"
>   ttl     = 60
>   type    = "CNAME"
>   records = ["${aws_acm_certificate.nexus_certificate.domain_validation_options.0.resource_record_value}"]
> }
>
> If anybody has any tips or advice, I'd very much appreciate it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-providers/terraform-provider-aws/pull/2813#issuecomment-369315950>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAETogsexXn940uyq45yJVl-47XmzbqAks5tZYzJgaJpZM4RP30U>
> .
>
